### PR TITLE
Add HostDialer interface

### PR DESCRIPTION
**Pull request by martin-sucha**
There are multiple use cases where the Dialer interface is not
sufficient.

When using TLS, users need to control also TLS context per host.
For example, host certificates might be either UUID in common name,
some hostname, IP address per host, etc.

Hosted instances or instances deployed in Kubernetes cluster tend
to be behind proxies. A proxy might use TLS server name indication
to identify which database host to connect to.

This commit adds Scylla-specific ShardDialer interface,
so that HostDialer works with shards as well.

### CODE REVIEW

**Comment by martin-sucha** [comment]
It is not clear to me why the integration test failed.

**Comment by martin-sucha** [comment]
This is a Scylla-specific version of pull request https://github.com/scylladb/gocql/pull/98 (posting so that those two are linked).

**Comment by martin-sucha** [comment]
The integration test works for me when running locally.

**Comment by zimnx** [comment]
This test was failing for me on my PR when driver didn't use shard aware ports to establish connections. Another retry on local machine happened so eventually pool grow using non-shard aware. On slower machine it must happen in the first try.
It's not great, I know.

Enable debug logs and check if in this test connections are using shard-aware port.

**Comment by zimnx** [comment]
Try updating Scylla to 4.6.3, your commits works fine with my PR #97 where I updated it, as well expclitly set shard_aware ports.

File: scylla.go [hunk 1]
```diff
@@ -484,53 +488,128 @@ func (p *scyllaConnPicker) GetCustomDialer() Dialer {
...
+// ShardDialer is like HostDialer but is shard-aware.
+// If the driver wants to connect to a specific shard, it will call DialShard,
+// otherwise it will call DialHost.
+type ShardDialer interface {
+	HostDialer
+
+	// DialShard establishes a connection to the specified shard ID out of nrShards.
+	// The returned connection must be directly usable for CQL protocol,
+	// specifically DialShard is responsible also for setting up the TLS session if needed.
+	DialShard(ctx context.Context, host *HostInfo, shardID, nrShards int) (DialedHost, error)
```
Comment (zimnx) [inline scylla.go:509]:
```suggestion
	DialShard(ctx context.Context, host *HostInfo, shardID, nrShards int) (*DialedHost, error)
```
It's easier to work with, and there's no need for copying.

**Review by zimnx — APPROVED** [review]
LGTM

File: cluster.go [hunk 2]
```diff
@@ -45,9 +45,12 @@ type ClusterConfig struct {
...
 	// versions the protocol selected is not defined (ie, it can be any of the supported in the cluster)
-	ProtoVersion       int
-	Timeout            time.Duration                            // connection timeout (default: 600ms)
-	ConnectTimeout     time.Duration                            // initial connection timeout, used during initial dial to server (default: 600ms)
+	ProtoVersion int
+	Timeout      time.Duration // connection timeout (default: 600ms)
+	// ConnectTimeout is initial connection timeout, used during initial dial to server.
+	// Default is 600ms.
+	// ConnectTimeout is used to set up the default dialer and is ignored if Dialer or HostDialer is provided.
+	ConnectTimeout     time.Duration
```
Comment (mmatczuk) [inline cluster.go:53]: Could we not reformat the code here that for the sake of consistency?

File: cluster.go [hunk 3]
```diff
@@ -58,13 +61,17 @@ type ClusterConfig struct {
...
+	// SocketKeepalive is used to set up the default dialer and is ignored if Dialer or HostDialer is provided.
+	SocketKeepalive   time.Duration
+	MaxPreparedStmts  int         // Sets the maximum cache size for prepared statements globally for gocql (default: 1000)
+	MaxRoutingKeyInfo int         // Sets the maximum cache size for query info about statements for each session (default: 1000)
+	PageSize          int         // Default page size to use for created sessions (default: 5000)
+	SerialConsistency Consistency // Sets the consistency for the serial part of queries, values can be either SERIAL or LOCAL_SERIAL (default: unset)
+	// SslOpts configures TLS use when HostDialer is not set.
+	// SslOpts is ignored if HostDialer is set.
+	SslOpts          *SslOptions
+	DefaultTimestamp bool // Sends a client side timestamp for all requests which overrides the timestamp at which it arrives at the server. (default: true, only enabled for protocol 3 and above)
```
Comment (mmatczuk) [inline cluster.go:74]: Again, maybe add a commit to promote the comments to doc strings and push it upstream?

File: conn.go [hunk 4]
```diff
@@ -205,23 +206,29 @@ func (s *Session) connect(ctx context.Context, host *HostInfo, errorHandler Conn
 	return s.dial(ctx, host, s.connCfg, errorHandler)
 }
 
-func (s *Session) connectWithDialer(ctx context.Context, host *HostInfo, errorHandler ConnErrorHandler, dialer Dialer) (*Conn, error) {
-	return s.dialWithDialer(ctx, host, s.connCfg, errorHandler, dialer)
+// connectShard establishes a connection to a shard.
+// if nrShards is zero, shard-aware port is disabled.
```
Comment (piodul) [inline conn.go:210]: Nitpick: rephrase this line as in the comment for `dialShard`, i.e. it's not that the shard-aware port itself is disabled, but rather dialing to that port.

File: scylla.go [hunk 5]
```diff
@@ -484,53 +488,128 @@ func (p *scyllaConnPicker) GetCustomDialer() Dialer {
...
+		if err := tconn.Handshake(); err != nil {
+			conn.Close()
+			return nil, err
+		}
+		conn = tconn
 	}
 
+	return &DialedHost{
+		Conn:            conn,
+		DisableCoalesce: sd.tlsConfig != nil, // write coalescing doesn't work with framing on top of TCP like in TLS.
```
Comment (mmatczuk) [inline scylla.go:592]: I do not understand what you mean here. Maybe it's a matter of the implementation but what exactly does not work?

File: conn.go (hunk 4 above)
Comment (mmatczuk) [inline conn.go:210]: Why not use capital letters when starting a new sentence?

File: cluster.go [hunk 6]
```diff
@@ -147,8 +154,14 @@ type ClusterConfig struct {
 
 	// Dialer will be used to establish all connections created for this Cluster.
 	// If not provided, a default dialer configured with ConnectTimeout will be used.
+	// Dialer is ignored if HostDialer is provided.
 	Dialer Dialer
 
+	// HostDialer will be used to establish all connections for this Cluster.
+	// To support shard-aware port, HostDialer should implement ShardDialer.
+	// If not provided, Dialer will be used instead.
+	HostDialer HostDialer
```
Comment (piodul) [inline cluster.go:163]: It's probably worth noting in the comment that a custom `HostDialer` must handle wrapping connection in TLS itself . There was no such requirement for `Dialer`, and the similar name might suggest that it's not required for `HostDialer` either.

File: conn.go [hunk 7]
```diff
@@ -454,6 +433,10 @@ func (s *startupCoordinator) options(ctx context.Context) error {
 	// Keep raw supported multimap for debug purposes
 	s.conn.supported = v.supported
 	s.conn.scyllaSupported = parseSupported(s.conn.supported)
+	s.conn.host.mu.Lock()
+	s.conn.host.scyllaShardAwarePort = s.conn.scyllaSupported.shardAwarePort
+	s.conn.host.scyllaShardAwarePortTLS = s.conn.scyllaSupported.shardAwarePortSSL
+	s.conn.host.mu.Unlock()
```
Comment (mmatczuk) [inline conn.go:439]: Refactor to method in host. Taking other object's mutex from outside is asking for trouble.

File: scylla.go [hunk 8]
```diff
@@ -484,53 +488,128 @@ func (p *scyllaConnPicker) GetCustomDialer() Dialer {
...
+
+	conn, err := sd.dialShardAware(ctx, addr, shardAwareAddress, iter)
+	if err != nil {
+		return nil, err
+	}
+
+	return sd.wrapTLS(conn, addr)
+}
+
+func (sd *scyllaDialer) wrapTLS(conn net.Conn, addr string) (*DialedHost, error) {
```
Comment (piodul) [inline scylla.go:579]: This function looks very similar to `tlsConfigForAddr`. Would it be possible to deduplicate the logic?

File: scylla.go (hunk 5 above)
Comment (martin-sucha) [reply scylla.go:592]: I copied this comment from the original code:

https://github.com/gocql/gocql/blob/89e6637b0c0191cb61ba4b9c1556d821ca55a8c7/connectionpool.go#L119

git blame points to the following explanation: https://github.com/gocql/gocql/commit/93aced8f230cbe935a771cf0918959b82b25003e

TLDR: TLS connnections can't use writev, so don't reduce the number of syscalls.

File: scylla.go (hunk 8 above)
Comment (martin-sucha) [reply scylla.go]: `wrapTLS` uses `tlsConfigForAddr`. Did you mean `defaultHostDialer.DialHost`? Anyway, I made wrapTLS a standalone function and used it in defaultHostDialer.DialHost as well.

File: cluster.go (hunk 6 above)
Comment (martin-sucha) [reply cluster.go:164]: Good point! It is documented in the HostDialer interface, but mentioning it here will be useful too. Added.

File: conn.go (hunk 4 above)
Comment (martin-sucha) [reply conn.go]: Updated. Thanks!

File: conn.go (hunk 7 above)
Comment (martin-sucha) [reply conn.go]: Good point. Moved. Thanks!

File: conn.go (hunk 4 above)
Comment (martin-sucha) [reply conn.go]: Fixed. Thanks!

File: cluster.go (hunk 3 above)
Comment (martin-sucha) [reply cluster.go:74]: Opened https://github.com/gocql/gocql/pull/1633 and https://github.com/scylladb/gocql/pull/100

**Comment by mmatczuk** [comment]
Please rebase.

File: scylla.go (hunk 5 above)
Comment (martin-sucha) [reply scylla.go]: I have updated the comment here and also for the HostDialer interface.

**Comment by martin-sucha** [comment]
Rebased. I have exported WrapTLS as discussed in https://github.com/gocql/gocql/pull/1629#discussion_r908525164

**Comment by martin-sucha** [comment]
I've noticed that WrapTLS did not use HandshakeContext, so I added context there.
